### PR TITLE
Part 3 of State Sync - Active and Crystallized Sync Services

### PR DESCRIPTION
**Pull request by terenc3t**
Last part of #347. This PR implements active and crystallized sync services and the tests for the services.

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -15,26 +15,30 @@ var log = logrus.WithField("prefix", "blockchain")
...
-	web3Service       *powchain.Web3Service
-	latestBeaconBlock chan *types.Block
-	processedHashes   [][32]byte
+	ctx                              context.Context
+	cancel                           context.CancelFunc
+	beaconDB                         *database.DB
+	chain                            *BeaconChain
+	web3Service                      *powchain.Web3Service
+	latestBeaconBlock                chan *types.Block
+	processedBlockHashes             [][32]byte
```
Comment (rawfalafel) [inline beacon-chain/blockchain/service.go]: Will these contain recently processed hashes, or every hash from genesis/last justified block? If the latter, will we load them all from disk at startup?

Comment (terenc3t) [reply beacon-chain/blockchain/service.go]: these hash -> block/state mappings are just temporary until we have a trie structure (binary merkle or sparse markle?) to store that permanently in disk

Comment (rawfalafel) [reply beacon-chain/blockchain/service.go]: gotcha :+1:

File: beacon-chain/sync/service.go [hunk 2]
```diff
@@ -25,35 +25,47 @@ var log = logrus.WithField("prefix", "sync")
...
+	ActiveStateHashBufferSize       int
+	ActiveStateBufferSize           int
+	CrystallizedStateHashBufferSize int
+	CrystallizedStateBufferSize     int
 }
 
 // DefaultConfig provides the default configuration for a sync service.
 func DefaultConfig() Config {
-	return Config{100, 100}
+	return Config{100, 100, 100, 100, 100, 100}
```
Comment (rawfalafel) [inline beacon-chain/sync/service.go]: Oh, I think I wrote this originally but we should make these named struct fields

Comment (terenc3t) [reply beacon-chain/sync/service.go]: Sure thing. I'll fix that before merge. Thanks for the reminder!

**Pull request merged**